src/imap/scan_folders.rs

```diff
@@ -0,0 +1,67 @@
+use std::time::{Instant, SystemTime};
+
+use crate::{config::Config, context::Context};
+use anyhow::Context as _;
+
+use crate::error::Result;
+use crate::imap::Imap;
+use async_std::prelude::*;
+
+impl Imap {
+    pub async fn scan_folders(&mut self, context: &Context) -> Result<()> {
+        warn!(context, "dbg starting scan");
+
+        // First of all, debounce to once per minute:
+        let mut last_scan = context.last_full_folder_scan.lock().await;
+        if let Some(last_scan) = *last_scan {
+            if last_scan.elapsed().as_secs() < 60 {
+                // For the first day after installation, we only debounce to 2s:
+                let configure = context.get_config(Config::ConfiguredTimestamp).await;
+                let configure = configure.context("scan_folders: not configured")?;
+                let configure: SystemTime = serde_json::from_str(&configure)?;
+
+                if configure.elapsed().unwrap().as_secs() > 24 * 60 * 60
+                    || last_scan.elapsed().as_secs() < 2
+                {
+                    warn!(context, "dbg not scanning, less than a minute elapsed");
+                    return Ok(());
+                }
+            }
+        }
+        last_scan.replace(Instant::now());
+
+        self.setup_handle(context).await?;
+        let session = self.session.as_mut();
+        let session = session.context("scan_folders(): IMAP No Connection established")?;
+        let folders: Vec<_> = session.list(Some(""), Some("*")).await?.collect().await;
```

Comment:
would be more efficient if you don't collect them, but rather process them in a `while let Some()` stream fashion

Reply:
I tried first but it's not possible due to borrowing problems:
```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
  --> src/imap/scan_folders.rs:43:13
   |
34 |         let session = self.session.as_mut();
   |                       ------------ first mutable borrow occurs here
...
43 |             self.fetch_new_messages(context, foldername, false).await?;
   |             ^^^^ second mutable borrow occurs here
...
66 |     }
   |     - first borrow might be used here, when `folders` is dropped and runs the destructor for type `impl std::marker::Send+futures_core::stream::Stream`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0499`.
error: could not compile `deltachat`.

To learn more, run the command again with --verbose.
```

Reply:
I would use self.session (mutably) to iterate over the folders and simultaneously use `self` (mutably) to fetch new messages.